src/Graph.cpp

```diff
@@ -234,21 +203,21 @@ void draw_links(Graph& graph, sf::RenderWindow& window)
 }
 
 
-void forward_propagate_beliefs(Graph& graph)
+void Graph::forward_propagate_beliefs()
 {
-    for (Link* link : graph.links)
+    for (Link* link : m_links)
     {
         link->to->update_beliefs(link->from);
         link->to->update_colour();
     }
 }
 
 
-void propagate_entities(Graph& graph, const time_period_t time_period)
+void Graph::propagate_entities(const time_period_t time_period)
 {
-    for (Link* link : graph.links)
+    for (Link* link : m_links)
     {
-        if (uniform_distribution_float(0, 1) < graph.spawn_chance)
+        if (uniform_distribution_float(0, 1) < m_spawn_chance)
         {
             EntityCircle* from_entity{ link->from };
```

Comment:
![image](https://user-images.githubusercontent.com/64105278/190349402-f9915008-9d65-4c99-8a4f-50a9ff1a5053.png)

Reply:
this is caused by a link being a null pointer in the m_links list

Reply:
![image](https://user-images.githubusercontent.com/64105278/190350190-f34771a4-c619-48e7-a89a-7c2b8d2ea183.png)

Reply:
presumably related to link_limit being 3000 and m_links being 3002 long??

Reply:
this loop iterates through m_links but also changes the size of m_links as it iterates, this is broken, make it only loop through as many links as exist in m_links _before_ the loop starts

Reply:
this causes issues when the resizing happens cos location in memory changes i think